rfd/0083-license-expiration-warnings.md

```diff
@@ -0,0 +1,135 @@
+```
+
+```
+$ build/tctl status -c /home/edward/teleport/teleport.yaml
+Cluster  localhost                                                               
+Version  10.1.2                                                                  
+host CA  never updated                                                           
+user CA  never updated                                                           
+db CA    never updated                                                           
+jwt CA   never updated                                                           
+CA pin   sha256:3d72102f020146d09ff400810f59f70b8163ebfe6ec1ecfb0b3b2a0c151592
+
+Your Teleport Enterprise Edition license has expired. Please reach out to [licenses@goteleport.com](mailto:licenses@goteleport.com) to obtain a new license. Inaction may lead to unplanned outage or degraded performance and support.
+```
+
+
+CLI does not support snoozing the warnings as you can in the webUI. Which provides the functionality to disable the warnings temporarily.
+
+### Web UI UX
+
+Web UI will display license expiration banner on top of the page with the following rules:
+Show warning (yellow) to users N days prior to expiration, 1 day snooze, based on license_warning_days role option with same warning text as the CLI
+Show error (red) to all users after expiration, no dismiss or snoozing available.
+Snoozing allows users to disable the warning in the web UI from being displayed for N days.
+
+### Configuration UX
+
+The time interval in which to show a warning before expiry can be configured on a per role basis.
+Interval can be configured by modifying a role option ‘license_warning_days’.
+Default will be 90 days, admins can configure if needed.
+Default preset roles will use the default. For example the editor/auditor/access roles.
+When multiple roles for the same user have different values for this interval the longest one will be used.
+This value cannot be configured to be below the minimum value of 10 days.
+
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: editor
+spec:
+  options:
+    license_warning_days: 90  # ← defaults to 30, min 10
+```
+
+Default values for default roles:
+Editor - 90 days
+Auditor - 30 days
+Access - 30 days
+
+If you have multiple roles, the longest time period takes precedence.
+
+## Implementation details
+
+Existing web api endpoint ‘/ping’ already returns a license warning. Due to this endpoint being unauthenticated and the proposed functionality requires the ability to differentiate based on identity/role this cannot be used as is.
+Instead this RFD proposes a new authenticated endpoint ‘/license’ be exposed to provide the required functionality. This would consist of
+- An auth api method under the hood that checks the license expiration and determines the warning to give.
+- A webapi endpoint that calls it.
+  - Method: GET
+  - Path: `/webapi/check-license`
```

Comment:
We probably don't need a separate endpoint now that we've decided to not do configurable intervals, and can just piggyback on the cluster alerts endpoint.

@hatched You were working on exposing cluster alerts API IIRC - is there a PR or a branch?

Reply:
The UI changes are in [this branch](https://github.com/gravitational/webapps/tree/hatched/banner) and I'm waiting on [the upgrade alerts pr](https://github.com/gravitational/teleport/pull/15774) to land and then I'll create a new pr with the routes and logic for the apiserver. I have it all working locally and can update this with PR's once ready.

Here is an example of the data structure that's being returned:

```json
{
  "alerts": [
    {
      "kind": "cluster_alert",
      "version": "v1",
      "metadata": {
        "name": "upgrade-suggestion",
        "labels": {
          "teleport.internal/alert-on-login": "yes",
          "teleport.internal/alert-permit-all": "yes"
        },
        "expires": "2022-08-31T17:26:05.728149Z"
      },
      "spec": {
        "severity": 5,
        "message": "A new major version of Teleport is available. Please consider upgrading your cluster to v10.",
        "created": "2022-08-30T17:26:05.728149Z"
      }
    }
  ]
}
```